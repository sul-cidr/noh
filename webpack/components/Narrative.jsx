import React, { Component } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import TabbedNarrative from "./TabbedNarrative";
import NonTabbedNarrative from "./NonTabbedNarrative";
import { setCurrentTime } from "../actionCreators";
import { getTime } from "../utils";

class Narrative extends Component {
  constructor(props) {
    super(props);

    this.state = {
      narrative: props.narrative,
      tabbed: this.checkIfTabbed()
    };

    this.narrativeOnClick = this.narrativeOnClick.bind(this);
  }

  componentWillMount() {
    this.checkIfTabbed(this.state.narrative, this.state.tabIndicator);
  }

  checkIfTabbed() {
    const re = /tabbed-narrative/;
    if (re.exec(this.props.narrative) != null) {
      return true;
    }
    return false;
  }

  narrativeOnClick(event) {
    if (event.target.tagName === "TIME") {
      this.props.updateStartTime(getTime(event.target));
    }
  }

  render() {
    let narrative = null;
    if (this.state.tabbed) {
      narrative = <TabbedNarrative narrative={this.state.narrative} />;
    } else {
      narrative = <NonTabbedNarrative narrative={this.state.narrative} />;
    }
    return (
      <div
        onClick={event => this.narrativeOnClick(event)}
        role="presentation"
        onKeyPress={null}
        className="narrative"
      >
        {narrative}
      </div>
    );
  }
}

Narrative.propTypes = {
  narrative: PropTypes.string,
  updateStartTime: PropTypes.func
};

Narrative.defaultProps = {
  narrative: "",
  updateStartTime: null
};

const mapDispatchToProps = dispatch => ({
  updateStartTime: time => dispatch(setCurrentTime(time))
});

export const Unwrapped = Narrative;
export default connect(null, mapDispatchToProps)(Narrative);
