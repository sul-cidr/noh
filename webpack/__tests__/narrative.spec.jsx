import React from "react";
import { Provider } from "react-redux";
import { shallow, mount } from "enzyme";
import configureMockStore from "redux-mock-store";
import Narrative, {
  Unwrapped as UnwrappedNarrative
} from "../components/Narrative";

const tabbedAnalysis =
  "<div title='section one with many words in title' class='tabbed-narrative'><p>Lorem ipsum dolor sit amet.</p></div><br><div title='section two'><p>Different stuff</p></div>";

const nonTabbedAnalysis =
  "Lorem ipsum dolor sit <time datetime='00:00:50'>amet</time>, <a datetime='00:00:50'>consectetur</a> adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam.";

describe("<UnwrappedNarrative>", () => {
  it("renders as expected without tabs", () => {
    const component = shallow(
      <UnwrappedNarrative narrative={nonTabbedAnalysis} />
    );
    expect(component).toMatchSnapshot();
  });

  it("renders as expected with tabs", () => {
    const component = shallow(
      <UnwrappedNarrative narrative={tabbedAnalysis} />
    );
    expect(component).toMatchSnapshot();
  });
});

describe("<Narrative>", () => {
  let wrapper;
  let store;

  beforeEach(() => {
    const initialState = { startTime: 10 };
    const mockStore = configureMockStore();

    store = mockStore(initialState);
    wrapper = mount(
      <Provider store={store}>
        <Narrative narrative={nonTabbedAnalysis} />
      </Provider>
    );
  });

  it("triggers the SET_CURRENT_TIME action with the right payload", () => {
    const action = { type: "SET_CURRENT_TIME", payload: 50 };
    wrapper.find("time").simulate("click");
    expect(store.getActions()[0]).toEqual(action);
  });

  it("does nothing when the element clicked is not a <time> element", () => {
    const action = { type: "SET_CURRENT_TIME", payload: 50 };
    wrapper.find("a").simulate("click");
    expect(store.getActions()[0]).not.toEqual(action);
  });
});
