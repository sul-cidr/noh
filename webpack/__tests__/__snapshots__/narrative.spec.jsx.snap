// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`<UnwrappedNarrative> renders as expected with tabs 1`] = `
<div
  className="narrative"
  onClick={[Function]}
  onKeyPress={null}
  role="presentation"
>
  <Connect(TabbedNarrative)
    narrative="<div title='section one with many words in title' class='tabbed-narrative'><p>Lorem ipsum dolor sit amet.</p></div><br /><div title='section two'><p>Different stuff</p></div>"
  />
</div>
`;

exports[`<UnwrappedNarrative> renders as expected without tabs 1`] = `
<div
  className="narrative"
  onClick={[Function]}
  onKeyPress={null}
  role="presentation"
>
  <NonTabbedNarrative
    narrative="Lorem ipsum dolor sit <time datetime='00:00:50'>amet</time>, <a datetime='00:00:50'>consectetur</a> adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. <h2 class='collapsible'>Header 2</h2>Ut enim ad minim veniam.<h2 class='collapsible collapsible-closed'>Header 3</h2>Et al."
  />
</div>
`;
